Changes to keep in line with new _validate_or_infer_batch_size for model tests.

cchoquette opened the pull request:
Changed:
models.py
models__test.py
to be in line with new implementations in latest tf2.0 version.

schien1729 commented:
Hi @cchoquette,

Thanks for the PR, which we've merged. In running the tests again, I noticed another issue. When running optimizers_test.py, I see this error:

```
TypeError: in converted code:
    relative to /build/work/a49763260a97a42ebcf878133f56f8a0bcbe/google3/runfiles/google3/third_party:

    py/tensorflow_privacy/privacy/bolt_on/optimizers_test.py:307 test_run  *
        pass
    py/tensorflow_privacy/privacy/bolt_on/optimizers.py:373 __exit__
        output_dim,
    py/tensorflow_privacy/privacy/bolt_on/optimizers.py:214 get_noise
        dtype=self.dtype
    tensorflow/python/ops/random_ops.py:473 random_gamma
        alpha = ops.convert_to_tensor(alpha, name="alpha", dtype=dtype)
    tensorflow/python/framework/ops.py:1129 convert_to_tensor
        return convert_to_tensor_v2(value, dtype, preferred_dtype, name)
    tensorflow/python/framework/ops.py:1187 convert_to_tensor_v2
        as_ref=False)
    tensorflow/python/framework/ops.py:1241 internal_convert_to_tensor
        ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
    tensorflow/python/framework/constant_op.py:338 _dimension_tensor_conversion_function
        raise TypeError("Cannot convert a TensorShape to dtype: %s" % dtype)

    TypeError: Cannot convert a TensorShape to dtype: <dtype: 'float32'>
```

The problem is that the value of `alpha` in [this line](https://github.com/tensorflow/privacy/blob/master/privacy/bolt_on/optimizers.py#L216) is a TensorShape. This can be fixed by adding `.value` at the [end of this line](https://github.com/tensorflow/privacy/blob/master/privacy/bolt_on/optimizers.py#L376), but this then causes models_test.py to break, since `layer.kernel.shape[0]` is already an int in that case.

Do you happen to know what's going on here? I imagine we can test for the type, but I wanted to see if this is a sign of an error somewhere else first.

Thanks!
Steve

cchoquette commented:
@schien1729 
Three points.
1. I am having a hard time testing to reproduce the same errors. I have switched from using tensorflow==2.0.0-beta1. Please advise if this is the best version to test on.
2. In the case you outlined above, I believe using .dims[0] would solve the problem. However, I cannot test this as I see different errors in beta1.
3. The errors I see are when calling BoltOnModel.fit (FitTests.test_fit, for all test cases). I have confirmed that the tutorial is broken due to the same error. The error is below:
`Traceback (most recent call last):
  File "/Users/christopherchoo/PycharmProjects/privacy/privacy/bolt_on/models.py", line 181, in fit
    **kwargs)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py", line 731, in fit
    use_multiprocessing=use_multiprocessing)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py", line 331, in fit
    total_epochs=epochs)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py", line 123, in run_one_epoch
    batch_outs = execution_function(iterator)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 86, in execution_function
    distributed_function(input_fn))
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py", line 461, in __call__
    return self._stateless_fn(*args, **kwds)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 1822, in __call__
    return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 1141, in _filtered_call
    self.captured_inputs)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 1224, in _call_flat
    ctx, args, cancellation_manager=cancellation_manager)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 511, in call
    ctx=ctx)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py", line 76, in quick_execute
    raise e
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py", line 61, in quick_execute
    num_outputs)
TypeError: An op outside of the function building code is being passed
a "Graph" tensor. It is possible to have Graph tensors
leak out of the function building context by including a
tf.init_scope in your function building code.
For example, the following function will fail:
  @tf.function
  def has_init_scope():
    my_constant = tf.constant(1.)
    with tf.init_scope():
      added = my_constant * 2
The graph tensor has name: cond/Identity:0
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/gen_math_ops.py", line 6688, in mul
    name, _ctx._post_execution_callbacks, x, y)
tensorflow.python.eager.core._FallbackException: This function does not handle the case of the path where all inputs are not already EagerTensors.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3296, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-24594aedbe52>", line 1, in <module>
    runfile('/Users/christopherchoo/PycharmProjects/privacy/tutorials/bolton_tutorial.py', wdir='/Users/christopherchoo/PycharmProjects/privacy/tutorials')
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/christopherchoo/PycharmProjects/privacy/tutorials/bolton_tutorial.py", line 98, in <module>
    epochs=2)
  File "/Users/christopherchoo/PycharmProjects/privacy/privacy/bolt_on/models.py", line 181, in fit
    **kwargs)
  File "/Users/christopherchoo/PycharmProjects/privacy/privacy/bolt_on/optimizers.py", line 374, in __exit__
    self.project_weights_to_r(True)
  File "/Users/christopherchoo/PycharmProjects/privacy/privacy/bolt_on/optimizers.py", line 172, in project_weights_to_r
    weight_norm = tf.norm(layer.kernel, axis=0)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/linalg_ops.py", line 493, in norm_v2
    name=name)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/linalg_ops.py", line 621, in norm
    tensor * math_ops.conj(tensor), axis, keepdims=True))
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/math_ops.py", line 899, in binary_op_wrapper
    return func(x, y, name=name)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/math_ops.py", line 1206, in _mul_dispatch
    return gen_math_ops.mul(x, y, name=name)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/gen_math_ops.py", line 6693, in mul
    x, y, name=name, ctx=_ctx)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/gen_math_ops.py", line 6730, in mul_eager_fallback
    ctx=_ctx, name=name)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py", line 76, in quick_execute
    raise e
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py", line 61, in quick_execute
    num_outputs)
TypeError: An op outside of the function building code is being passed
a "Graph" tensor. It is possible to have Graph tensors
leak out of the function building context by including a
tf.init_scope in your function building code.
For example, the following function will fail:
  @tf.function
  def has_init_scope():
    my_constant = tf.constant(1.)
    with tf.init_scope():
      added = my_constant * 2
The graph tensor has name: cond/Identity:0`

I am looking into this error and initially suspected it arose due to calling graph functions outside of the graph execution in .fit. However, commenting out lines 374-381 in optimizers.py: 
`self.project_weights_to_r(True)
    for layer in self.layers:
      input_dim = layer.kernel.shape[0]
      output_dim = layer.units
      noise = self.get_noise(input_dim,
                             output_dim,
                            )
      layer.kernel = tf.math.add(layer.kernel, noise)`
Leads to another mysterious error, this time located within fit.
`Traceback (most recent call last):
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3296, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-24594aedbe52>", line 1, in <module>
    runfile('/Users/christopherchoo/PycharmProjects/privacy/tutorials/bolton_tutorial.py', wdir='/Users/christopherchoo/PycharmProjects/privacy/tutorials')
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/Applications/PyCharm.app/Contents/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/christopherchoo/PycharmProjects/privacy/tutorials/bolton_tutorial.py", line 98, in <module>
    epochs=2)
  File "/Users/christopherchoo/PycharmProjects/privacy/privacy/bolt_on/models.py", line 181, in fit
    **kwargs)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py", line 731, in fit
    use_multiprocessing=use_multiprocessing)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py", line 331, in fit
    total_epochs=epochs)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py", line 123, in run_one_epoch
    batch_outs = execution_function(iterator)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py", line 86, in execution_function
    distributed_function(input_fn))
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py", line 461, in __call__
    return self._stateless_fn(*args, **kwds)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 1822, in __call__
    return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 1141, in _filtered_call
    self.captured_inputs)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 1224, in _call_flat
    ctx, args, cancellation_manager=cancellation_manager)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py", line 511, in call
    ctx=ctx)
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py", line 76, in quick_execute
    raise e
  File "/Users/christopherchoo/PycharmProjects/privacy/venv/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py", line 61, in quick_execute
    num_outputs)
TypeError: An op outside of the function building code is being passed
a "Graph" tensor. It is possible to have Graph tensors
leak out of the function building context by including a
tf.init_scope in your function building code.
For example, the following function will fail:
  @tf.function
  def has_init_scope():
    my_constant = tf.constant(1.)
    with tf.init_scope():
      added = my_constant * 2
The graph tensor has name: cond/Identity:0`

I am looking into the errors; if you have any advice on the reason these are occurring it would be very helpful.

cchoquette commented:
@schien1729 @npapernot 
Updating the bug:
I have confirmed the issue lies with the project_weights_to_r function inside optimizers.BoltOn, which is creating the ops responsible for this error.
I am not sure yet as to how to solve this issue.
